opa eval: panic when given bad input for --unknowns

`opa eval --help` says

```
-u, --unknowns stringArray                                      set paths to treat as unknown during partial evaluation (default [input])
```

so it's not completely out of the question to try setting it to `[input,data.posts]`. When you get then is a panic:

```
% opa eval --partial -d q.rego -fpretty -u '[input,data.posts]' data.example.allow
panic: illegal value

goroutine 1 [running]:
github.com/open-policy-agent/opa/topdown.newSaveSetElem(...)
	/Users/stephan/Sources/open-policy-agent/opa/topdown/save.go:132
github.com/open-policy-agent/opa/topdown.(*saveSet).Push(0xc0000177a0, {0xc000135c20, 0x1, 0x100071205?}, 0xc000020180)
	/Users/stephan/Sources/open-policy-agent/opa/topdown/save.go:31 +0x434
github.com/open-policy-agent/opa/topdown.newSaveSet({0xc000135c20, 0x1, 0x1}, 0xc000020180, 0x0)
	/Users/stephan/Sources/open-policy-agent/opa/topdown/save.go:26 +0xbb
github.com/open-policy-agent/opa/topdown.(*Query).PartialRun(0xc000425180, {0x101f32350, 0x102a8b840})
	/Users/stephan/Sources/open-policy-agent/opa/topdown/query.go:370 +0x72e
github.com/open-policy-agent/opa/rego.(*Rego).partial(0xc0003c6e08, {0x101f32350, 0x102a8b840}, 0xc000365860)
	/Users/stephan/Sources/open-policy-agent/opa/rego/rego.go:2455 +0x93c
github.com/open-policy-agent/opa/rego.PreparedPartialQuery.Partial({{0xc0003c6e08?, 0xc0004034d0?}}, {0x101f32350, 0x102a8b840}, {0xc000445c90?, 0xc0000d98c0?, 0x1035df108?})
	/Users/stephan/Sources/open-policy-agent/opa/rego/rego.go:507 +0x149
github.com/open-policy-agent/opa/cmd.evalOnce({0x101f32350, 0x102a8b840}, 0xc0001bd200)
	/Users/stephan/Sources/open-policy-agent/opa/cmd/eval.go:472 +0xf07
github.com/open-policy-agent/opa/cmd.eval({_, _, _}, {0xc000402a50, 0x0, 0x1, {0xc000444e30, 0x1, 0x1}, {0x102a8b840, ...}, ...}, ...)
	/Users/stephan/Sources/open-policy-agent/opa/cmd/eval.go:374 +0x3f0
github.com/open-policy-agent/opa/cmd.init.5.func2(0xc0003f2f08?, {0xc000376380?, 0x4?, 0x1019e5a41?})
	/Users/stephan/Sources/open-policy-agent/opa/cmd/eval.go:292 +0x6e
github.com/spf13/cobra.(*Command).execute(0xc0003f2f08, {0xc000376310, 0x7, 0x7})
	/Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:989 +0xa91
github.com/spf13/cobra.(*Command).ExecuteC(0x102a4cc80)
	/Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:1117 +0x3ff
github.com/spf13/cobra.(*Command).Execute(0xc0000061c0?)
	/Users/stephan/Sources/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:1041 +0x13
main.main()
	/Users/stephan/Sources/open-policy-agent/opa/main.go:14 +0x1a
```

It would be nicer if we told the user that their argument to `--unknowns` is of the wrong type.

(BTW the right format would be `--unknowns input --unknowns data.posts`)

--- a/cmd/eval.go
+++ b/cmd/eval.go
@@ -137,7 +137,7 @@ func validateEvalParams(p *evalCommandParams, cmdArgs []string) error {
 		return errors.New("invalid output format for evaluation")
 	}
 
-	//* check if illegal arguments is passed with unknowns flag
+	// check if illegal arguments is passed with unknowns flag
 	for _, unknwn := range p.unknowns {
 		term, err := ast.ParseTerm(unknwn)
 		if err != nil {
@@ -148,7 +148,7 @@ func validateEvalParams(p *evalCommandParams, cmdArgs []string) error {
 		case ast.Ref:
 			return nil
 		default:
-			return errors.New(errIllegalUnknownsArg.Error())
+			return errIllegalUnknownsArg
 		}
 	}
 

--- a/cmd/eval.go
+++ b/cmd/eval.go
@@ -11,7 +11,6 @@ import (
 	"fmt"
 	"io"
 	"os"
-	"regexp"
 	"strconv"
 	"strings"
 	"time"
@@ -139,9 +138,16 @@ func validateEvalParams(p *evalCommandParams, cmdArgs []string) error {
 	}
 
 	//* check if illegal arguments is passed with unknowns flag
-	regexArr := regexp.MustCompile(`^\[.*\]$`)
 	for _, unknwn := range p.unknowns {
-		if regexArr.MatchString(unknwn) {
+		term, err := ast.ParseTerm(unknwn)
+		if err != nil {
+			return err
+		}
+
+		switch term.Value.(type) {
+		case ast.Ref:
+			return nil
+		default:
 			return errors.New(errIllegalUnknownsArg.Error())
 		}
 	}

--- a/cmd/eval.go
+++ b/cmd/eval.go
@@ -11,6 +11,7 @@ import (
 	"fmt"
 	"io"
 	"os"
+	"regexp"
 	"strconv"
 	"strings"
 	"time"
@@ -33,6 +34,10 @@ import (
 	"github.com/open-policy-agent/opa/topdown"
 	"github.com/open-policy-agent/opa/topdown/lineage"
 	"github.com/open-policy-agent/opa/util"
+)
+
+var (
+	illegalUnknownsArg = errors.New("illegal argument with --unknowns, specify string with one or more --unknowns")
 )
 
 type evalCommandParams struct {
@@ -109,6 +114,7 @@ func newEvalCommandParams() evalCommandParams {
 }
 
 func validateEvalParams(p *evalCommandParams, cmdArgs []string) error {
+
 	if len(cmdArgs) > 0 && p.stdin {
 		return errors.New("specify query argument or --stdin but not both")
 	} else if len(cmdArgs) == 0 && !p.stdin {
@@ -130,6 +136,14 @@ func validateEvalParams(p *evalCommandParams, cmdArgs []string) error {
 		return errors.New("invalid output format for partial evaluation")
 	} else if !p.partial && of == evalSourceOutput {
 		return errors.New("invalid output format for evaluation")
+	}
+
+	//* check if illegal arguments is passed with unknowns flag
+	regexArr := regexp.MustCompile(`^\[.*\]$`)
+	for _, unknwn := range p.unknowns {
+		if regexArr.MatchString(unknwn) {
+			return errors.New(illegalUnknownsArg.Error())
+		}
 	}
 
 	if p.optimizationLevel > 0 {

--- a/cmd/eval.go
+++ b/cmd/eval.go
@@ -37,7 +37,7 @@ import (
 )
 
 var (
-	illegalUnknownsArg = errors.New("illegal argument with --unknowns, specify string with one or more --unknowns")
+	errIllegalUnknownsArg = errors.New("illegal argument with --unknowns, specify string with one or more --unknowns")
 )
 
 type evalCommandParams struct {
@@ -142,7 +142,7 @@ func validateEvalParams(p *evalCommandParams, cmdArgs []string) error {
 	regexArr := regexp.MustCompile(`^\[.*\]$`)
 	for _, unknwn := range p.unknowns {
 		if regexArr.MatchString(unknwn) {
-			return errors.New(illegalUnknownsArg.Error())
+			return errors.New(errIllegalUnknownsArg.Error())
 		}
 	}
 

--- a/cmd/eval_test.go
+++ b/cmd/eval_test.go
@@ -30,16 +30,47 @@ import (
 )
 
 func TestEvalWithIllegalUnknownArgs(t *testing.T) {
-	params := newEvalCommandParams()
-	params.unknowns = []string{"[input, data.posts]"}
-	params.partial = true
 
-	err := validateEvalParams(&params, []string{"data"})
-
-	if !strings.EqualFold(err.Error(), errIllegalUnknownsArg.Error()) {
-		t.Errorf("expected %s; got %s", errIllegalUnknownsArg.Error(), err.Error())
+	tests := []struct {
+		name        string
+		unknowns    string
+		expectedErr error
+	}{
+		{
+			name:        "happy path: passing ref as unknown",
+			unknowns:    "data.posts",
+			expectedErr: nil,
+		},
+		{
+			name:        "passing ; separated values",
+			unknowns:    "input; data.posts",
+			expectedErr: errors.New("expected exactly one term but got: input; data.posts"),
+		},
+		{
+			name:        "passing array as unknown",
+			unknowns:    "[input, data.posts]",
+			expectedErr: errIllegalUnknownsArg,
+		},
+		{
+			name:        "passing set as unknown",
+			unknowns:    "{input, data.posts}",
+			expectedErr: errIllegalUnknownsArg,
+		},
 	}
 
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			params := newEvalCommandParams()
+			params.unknowns = []string{tt.unknowns}
+			params.partial = true
+
+			err := validateEvalParams(&params, []string{"data"})
+
+			if tt.expectedErr != nil && !strings.EqualFold(err.Error(), tt.expectedErr.Error()) {
+				t.Errorf("expected %s; got %s", errIllegalUnknownsArg.Error(), err.Error())
+			}
+		})
+	}
 }
 
 func TestEvalExitCode(t *testing.T) {

--- a/cmd/eval_test.go
+++ b/cmd/eval_test.go
@@ -37,14 +37,19 @@ func TestEvalWithIllegalUnknownArgs(t *testing.T) {
 		expectedErr error
 	}{
 		{
-			name:        "happy path: passing ref as unknown",
-			unknowns:    "data.posts",
+			name:        "happy path: passing input ref as unknown",
+			unknowns:    "input",
 			expectedErr: nil,
 		},
 		{
-			name:        "passing ; separated values",
-			unknowns:    "input; data.posts",
-			expectedErr: errors.New("expected exactly one term but got: input; data.posts"),
+			name:        "happy path: passing input.users ref as unknown",
+			unknowns:    "input.users",
+			expectedErr: nil,
+		},
+		{
+			name:        "passing multiple refs with ; separated",
+			unknowns:    "input;input.users",
+			expectedErr: errors.New("expected exactly one term but got: input; input.users"),
 		},
 		{
 			name:        "passing array as unknown",

--- a/cmd/eval_test.go
+++ b/cmd/eval_test.go
@@ -36,8 +36,8 @@ func TestEvalWithIllegalUnknownArgs(t *testing.T) {
 
 	err := validateEvalParams(&params, []string{"data"})
 
-	if !strings.EqualFold(err.Error(), illegalUnknownsArg.Error()) {
-		t.Errorf("expected %s; got %s", illegalUnknownsArg.Error(), err.Error())
+	if !strings.EqualFold(err.Error(), errIllegalUnknownsArg.Error()) {
+		t.Errorf("expected %s; got %s", errIllegalUnknownsArg.Error(), err.Error())
 	}
 
 }

--- a/cmd/eval_test.go
+++ b/cmd/eval_test.go
@@ -29,6 +29,19 @@ import (
 	"github.com/open-policy-agent/opa/util/test"
 )
 
+func TestEvalWithIllegalUnknownArgs(t *testing.T) {
+	params := newEvalCommandParams()
+	params.unknowns = []string{"[input, data.posts]"}
+	params.partial = true
+
+	err := validateEvalParams(&params, []string{"data"})
+
+	if !strings.EqualFold(err.Error(), illegalUnknownsArg.Error()) {
+		t.Errorf("expected %s; got %s", illegalUnknownsArg.Error(), err.Error())
+	}
+
+}
+
 func TestEvalExitCode(t *testing.T) {
 	params := newEvalCommandParams()
 	params.fail = true
@@ -764,8 +777,8 @@ func TestEvalWithSchemaFileWithRemoteRef(t *testing.T) {
 		"p.rego": `package p
 import rego.v1
 
-r if { 
-	input.metadata.clusterName == "NAME" 
+r if {
+	input.metadata.clusterName == "NAME"
 }`,
 	}
 
@@ -1296,7 +1309,7 @@ p if {
 	y := 2
 	z := 3
 	x == z - y
-} 
+}
 `,
 			},
 			expected: `%SKIP_LINE%
@@ -1335,7 +1348,7 @@ p if {
 	y := 2
 	z := 3
 	x == z - y
-} 
+}
 `,
 			},
 			expected: `%SKIP_LINE%
@@ -1380,7 +1393,7 @@ p if {
 	x := v
 
 	x.foo[_] == "a"
-} 
+}
 `,
 			},
 			expected: `%SKIP_LINE%

